doc: nRF541l15: Add information on ECIES-X25519 support (#22056)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54l/ecies_x25519.rst

.. _ecies_x25519:

> MCUboot AES Image Encryption with ECIES-X25519 key exchange

Author: uppercase `key exchange`

###########################################################

.. contents::
   :local:
   :depth: 2

Introduction
************

> MCUboot, when used with nRF54l15, can be configured to support encrypted images using AES.

Reviewer: ```suggestion
MCUboot on the nRF54L15 SoC can support encrypted images using AES.
```

> Images are encrypted using AES and ECIES-X25519 is used for key delivery (exchange) within image.
> When image encryption is enabled user may choose to either upload just signed or encrypted images to be swapped during boot.
> If MCUboot finds encrypted image in secondary slot it will decrypt it while swapping slots.
> Image that has been encrypted before being swapped into primary slot will be re-encrypted in case when swap is reverted.

Reviewer: ```suggestion
Images are encrypted using AES, and ECIES-X25519 is used for key delivery (exchange) within the image.
When image encryption is enabled, you can choose to upload either signed images or encrypted images to be swapped during boot.
If MCUboot finds an encrypted image in the secondary slot, it decrypts the image during the slot swap process.
An image that was encrypted before being swapped into the primary slot is re-encrypted if the swap is later reverted.
```

Limitations
***********

ECIES-X25519 key exchange is supported only when ED25519 signature is selected, which is default for nRF54l15.
Encryption is not supported with MCUboot Direct-XIP modes.
Storing ECIES-X25519 device private key in KMU is not supported at this moment.

> SHA256 is currently used by HMAC/HKDF, which is subject to change.

Author: ```suggestion
SHA256 is currently used by HMAC/HKDF, which is a subject to change.
```

> ECIES-X25519 key exchange is supported only when ED25519 signature is selected, which is default for nRF54l15.
> Encryption is not supported with MCUboot Direct-XIP modes.
> Storing ECIES-X25519 device private key in KMU is not supported at this moment.
> SHA256 is currently used by HMAC/HKDF, which is a subject to change.

Reviewer: ```suggestion
The current implementation has the following limitations:

* On the nRF54L15 SoC, ECIES-X25519 key exchange is supported only when the ED25519 signature algorithm is selected.
  This is the default configuration.
* Encryption is not supported when using MCUboot in Direct-XIP mode.
* Storing the ECIES-X25519 device private key in the Key Management Unit (KMU) is not supported at this time.
* SHA-256 is currently used by HMAC and HKDF, but this may change in future releases.
```

HMAC/HKDF impact on TLV and Key Exchange
****************************************

Encrypted image contains TLV that provides public key for ECIES-X25519 key exchange, encrypted AES key and MAC tag of the encrypted key.

> Derivation of ASE key encryption key is involves HKDF and MAC tag is generated using HMAC; both HKDF and HMAC at this point use SHA256.

Reviewer 2: ```suggestion
Derivation of AES key encryption key is involves HKDF and MAC tag is generated using HMAC; both HKDF and HMAC at this point use SHA256.
```

> Derivation of ASE key encryption key is involves HKDF and MAC tag is generated using HMAC; both HKDF and HMAC at this point use SHA256.

Reviewer 2: ```suggestion
Derivation of ASE key encryption key is involves HKDF and MAC tag generated using HMAC; both HKDF and HMAC at this point use SHA256.
```

> Derivation of ASE key encryption key is involves HKDF and MAC tag is generated using HMAC; both HKDF and HMAC at this point use SHA256.

Author: Rephrased.

> Encrypted image contains TLV that provides public key for ECIES-X25519 key exchange, encrypted AES key and MAC tag of the encrypted key.
> Derivation of AES key encryption key involves HKDF, and MAC tag, of encryption key, is generated using HMAC; both HKDF and HMAC use sha256, at this point.
> While using SHA256 is not a problem and does not significantly impact security it increases code size, as it requires addition of the algorithm support together with SHA512, which is already used by ED25519.
> There might be an option, in the future, to switch to using SHA512, but pre-installed MCUboot instance will not be able to boot images signed with TLV using SHA512.

Reviewer: ```suggestion
An encrypted image includes a TLV that contains the public key used for ECIES-X25519 key exchange, the encrypted AES key, and the MAC tag of the encrypted key.
The key used to encrypt the AES key is derived using HKDF, and the MAC tag is generated using HMAC.
Both HKDF and HMAC currently use the SHA-256 hash algorithm.

While the use of SHA-256 does not pose a security concern and has minimal impact on performance, it increases code size because SHA-256 support must be included alongside SHA-512, which is already used by the ED25519 signature algorithm.

Pre-installed MCUboot instances will not be able to boot images that use TLVs generated with different hash  algorithms that might be supported by this implementation in the future.
```

Building Application with Image Encryption
******************************************

> Building application with MCUboot that has image encryption enabled is performed with command:

Reviewer: ```suggestion
To build an application that uses MCUboot with image encryption enabled, do the following:
```

Reviewer: Where are the steps for VS code?

.. parsed-literal::
   :class: highlight

    west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y

> The :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` enables MCUboot encryption support and key exchange method is selected by type of selected key for signature; in case of nRF54115 ED25519 signature is selected by default.

Author: ```suggestion
The :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` enables MCUboot encryption support and key exchange method is selected by type of selected key for signature; in case of nRF54l15 ED25519 signature is selected by default.
```

> The :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` enables MCUboot encryption support and key exchange method is selected by type of selected key for signature; in case of nRF54l15 ED25519 signature is selected by default.
> The encrypted image will be stored, within application build directory, under name zephyr.signed.encrypted.bin and zephyr.signed.encrypted.hex, where the former represents binary image that can be used with MCUmgr for DFU.

Reviewer: ```suggestion
The :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` option enables encryption support in MCUboot.

The key exchange method is determined by the type of signature key selected.
For the nRF54L15 SoC, the ED25519 signature algorithm is selected by default.

When encryption is enabled, the encrypted image is generated in the application build directory under the names :file:`zephyr.signed.encrypted.bin` and :file:`zephyr.signed.encrypted.hex`.

The BIN file is a binary image suitable for DFU operations using MCUmgr.
```

> With additional option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` user should pass ECIES-X25519 private key in pem format; the key will be built into MCUboot.

Reviewer: ```suggestion
When the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option is enabled, you must provide an ECIES-X25519 private key in PEM format.
This key is built into the MCUboot image during the build process.
See the following example:
```

Reviewer: Same here, VScode steps?

.. parsed-literal::
   :class: highlight

   west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y -DSB_CONFIG_BOOT_ENCRYPTION_KEY_FILE=\"<path to key.pem>\"

> Note that public key derived from the specified file will be used to sign an application image.

Author: (imgtool note?)

> Note that public key derived from the specified file will be used to sign an application image.

Author: ```suggestion
Note that public encryption key, derived from the specified file, will be used to sign an application image.
```

> Note that public encryption key, derived from the specified file, will be used to sign an application image. This public key is automatically derived from private key by imgtool, which is invoked by build system when signing an image.

Author: Rephrased entire sentence.

> Note that public key, derived from the specified file, will be added to image and later used, on a device, to derive the decryption key for an application image. This public key is automatically derived from private key by imgtool, which is invoked by build system when signing an image.

Reviewer: ```suggestion
.. note::

*  The public key, derived from the specified private key file, is added to the image and later used on the device to derive the decryption key for the application image.

   This public key is automatically derived from the private key by ``imgtool``, which is invoked by the build system when signing the image.

* The Sysbuild option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets the MCUboot configuration option :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`.

   Similarly, the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option sets both :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE` for MCUboot and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for the default application.

   These values are then passed to ``imgtool`` for encrypting the application image.

   You cannot override these options using MCUboot or application-level Kconfig options, as they are enforced by Sysbuild.

```

.. note::

>    Stsbuild options :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`;

Author: Fixed

>    :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`;
>    :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE`
>    and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for default application, which is then passed to imagetool for application image encryption.

Author: ```suggestion
   Stsbuild options :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`;
   :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE`
   and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for default application, which is then passed to imagetool for application image encryption. You can not override, options enforced by sysbuild, with MCUboot nor application Kconfig options to correspond to them.
```

> .. note::
> 
>    Sybuild options :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`;
>    :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` sets MCUboot :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE`
>    and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for default application, which is then passed to imgtool for application image encryption. You can not override, options enforced by sysbuild, with MCUboot nor application Kconfig options to correspond to them.

Reviewer: ```suggestion
```

.. _ug_nrf54l_ecies_x25519:

> MCUboot AES Image Encryption with ECIES-X25519 Key Exchange

Reviewer 3: Noted - thanks, let's only fix the letter casing then: 
``MCUboot AES image encryption with ECIES-X25519 key exchange``

###########################################################

.. contents::
   :local:
   :depth: 2

MCUboot on the nRF54L15 SoC can support encrypted images using AES.
Images are encrypted using AES, and ECIES-X25519 is used for key delivery (exchange) within the image.
When image encryption is enabled, you can choose to upload either signed images or encrypted images to be swapped during boot.
If MCUboot finds an encrypted image in the secondary slot, it decrypts the image during the slot swap process.
An image that was encrypted before being swapped into the primary slot is re-encrypted if the swap is later reverted.

Limitations
***********

The current implementation has the following limitations:

* On the nRF54L15 SoC, ECIES-X25519 key exchange is supported only when the ED25519 signature algorithm is selected.
  This is the default configuration.

> * Encryption is not supported when using MCUboot in Direct-XIP mode.

Reviewer 3: ```suggestion
* Encryption is not supported when using MCUboot in direct-xip mode.
```

> * Storing the ECIES-X25519 device private key in the Key Management Unit (KMU) is not supported at this time.

Reviewer 3: ```suggestion
* Storing the ECIES-X25519 device private key in the Key Management Unit (KMU) is currently not supported.
```

> * SHA-256 is currently used by HMAC and HKDF, but this may change in future releases.

Reviewer 3: ```suggestion
* HMAC and HKDF tools currently use the SHA-256 hash algorithm.
```

HMAC/HKDF impact on TLV and Key Exchange
****************************************

An encrypted image includes a TLV that contains the public key used for ECIES-X25519 key exchange, the encrypted AES key, and the MAC tag of the encrypted key.
The key used to encrypt the AES key is derived using HKDF, and the MAC tag is generated using HMAC.

> Both HKDF and HMAC currently use the SHA-256 hash algorithm.

Reviewer 3: Isn't this repetition of line 26? Do we need to have it in both places?

> While the use of SHA-256 does not pose a security concern and has minimal impact on performance, it increases code size because SHA-256 support must be included alongside SHA-512, which is already used by the ED25519 signature algorithm.

Reviewer 3: ```suggestion
While the use of SHA-256 does not pose a security concern and has a minimal impact on performance, it increases the code size. 
This is because SHA-256 support must be included in addition to SHA-512, which is already used by the ED25519 signature algorithm.
```

> Pre-installed MCUboot instances will not be able to boot images that use TLVs generated with different hash  algorithms that might be supported by this implementation in the future.

Reviewer 3: General note: not sure if we should declare what might be supported in the future. 
```suggestion
Additionally, pre-installed MCUboot instances will not be able to boot images that use TLVs generated with different hash algorithms.
```

Building Application with Image Encryption
******************************************

> To build an application that uses MCUboot with image encryption enabled, do the following:

Reviewer 3: ```suggestion
To build an application that uses MCUboot with image encryption enabled, run the following command:
```

.. parsed-literal::
   :class: highlight

    west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y

The :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` option enables encryption support in MCUboot.

The key exchange method is determined by the type of signature key selected.
For the nRF54L15 SoC, the ED25519 signature algorithm is selected by default.

When encryption is enabled, the encrypted image is generated in the application build directory under the names :file:`zephyr.signed.encrypted.bin` and :file:`zephyr.signed.encrypted.hex`.

> The BIN file is a binary image suitable for DFU operations using MCUmgr.

Reviewer 3: ```suggestion
The BIN file is a binary image suitable for DFU operations using :ref:`MCUmgr<dfu_tools_mcumgr_cli>`.
```

When the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option is enabled, you must provide an ECIES-X25519 private key in PEM format.
This key is built into the MCUboot image during the build process.
See the following example:

.. parsed-literal::
   :class: highlight

   west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y -DSB_CONFIG_BOOT_ENCRYPTION_KEY_FILE=\"<path to key.pem>\"

.. note::
   * The public key, derived from the specified private key file, is added to the image and later used on the device to derive the decryption key for the application image.
     This public key is automatically derived from the private key by ``imgtool``, which is invoked by the build system when signing the image.

>    * The Sysbuild option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets the MCUboot configuration option :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`.

Reviewer 3: ```suggestion
   * The sysbuild option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets the MCUboot configuration option :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`.
```

     Similarly, the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option sets both :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE` for MCUboot and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for the default application.

   These values are then passed to ``imgtool`` for encrypting the application image.

>    You cannot override these options using MCUboot or application-level Kconfig options, as they are enforced by Sysbuild.

Reviewer 3: ```suggestion
   You cannot override these options using MCUboot or application-level Kconfig options, as they are enforced by sysbuild.
```

> .. _ug_nrf54l_ecies_x25519:

Reviewer 3: In addition to VS code instructions, please include the reference to this page in the following file and line:
https://github.com/nrfconnect/sdk-nrf/blob/24d63eb67e72847e6b7d2ef19502857aab3219c8/doc/nrf/app_dev/device_guides/nrf54l/dfu_config.rst?plain=1#L137

I.e.:

``For detailed information on ECIES-X25519 support, refer to the :ref:`ug_nrf54l_ecies_x25519` documentation page.``

   :class: highlight

   west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y -DSB_CONFIG_BOOT_ENCRYPTION_KEY_FILE=\"<path to key.pem>\"

.. note::
   * The public key, derived from the specified private key file, is added to the image and later used on the device to derive the decryption key for the application image.
     This public key is automatically derived from the private key by ``imgtool``, which is invoked by the build system when signing the image.
   * The sysbuild option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets the MCUboot configuration option :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`.
     Similarly, the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option sets both :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE` for MCUboot and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for the default application.

   These values are then passed to ``imgtool`` for encrypting the application image.

   You cannot override these options using MCUboot or application-level Kconfig options, as they are enforced by sysbuild.

Enabling encryption in |nRFVSC| projects
****************************************

> Generally, enabling encryption in projects created using |nRFVSC| requires passing Kconfig options, mentioned in above paragraphs, to project configuration. This action is best described in :ref:`cmake_options`.

Reviewer 3: I don't think these are solid instructions that the user could follow. It says "generally, it works like this", which almost sounds like it wasn't checked

   :class: highlight

   west build -b *board_target* -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DSB_CONFIG_BOOT_ENCRYPTION=y -DSB_CONFIG_BOOT_ENCRYPTION_KEY_FILE=\"<path to key.pem>\"

.. note::
   * The public key, derived from the specified private key file, is added to the image and later used on the device to derive the decryption key for the application image.
     This public key is automatically derived from the private key by ``imgtool``, which is invoked by the build system when signing the image.
   * The sysbuild option :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` sets the MCUboot configuration option :kconfig:option:`CONFIG_BOOT_ENCRYPT_IMAGE`.
     Similarly, the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` option sets both :kconfig:option:`CONFIG_BOOT_ENCRYPTION_KEY_FILE` for MCUboot and :kconfig:option:`CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE` for the default application.

   These values are then passed to ``imgtool`` for encrypting the application image.

   You cannot override these options using MCUboot or application-level Kconfig options, as they are enforced by sysbuild.

Enabling encryption in |nRFVSC| projects
****************************************

> You will need to get familiar with `How to work with build configurations`_ page to be able to setup encryption in |nRFVSC|.
> You will have to add the :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` and :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` using Extra CMake arguments.
> In case when already existing project is being modified, it has to be re-generated, otherwise the new settings may not take effect.

Reviewer 3: Shouldn't this section appear before "Building application with image encryption"?

```suggestion
To correctly set up encryption in |nRFVSC|, you must familiarize yourself with `How to work with build configurations`_.
When configuring build options, ensure to include :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION` and :kconfig:option:`SB_CONFIG_BOOT_ENCRYPTION_KEY_FILE` Kconfig options using extra CMake arguments.

If you are modifying an existing project, you must regenerate it to activate new settings.
```

.. _ug_nrf54l_ecies_x25519:

MCUboot AES image encryption with ECIES-X25519 key exchange
###########################################################

.. contents::
   :local:
   :depth: 2

> 

Author: ```suggestion

```
